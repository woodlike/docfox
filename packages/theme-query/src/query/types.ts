import { SxStyleProp } from 'theme-ui';

export interface ThemeQueryConfig {
  theme: ThemeProps;
  styles: StylesType;
}

export interface ThemeProps {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [x: string]: any;
}
export type StylesType = 'object' | 'templateLitereal';
export type QueryResult = ((idx: number | 'all') => SxStyleProp) & SxStyleProp;
export type ThemeQuery = (objPath: string) => QueryResult;
