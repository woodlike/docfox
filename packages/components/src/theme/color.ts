import { GenericThemeColor } from '.';

export const query = (
  colorQuery: string,
  color: GenericThemeColor,
  idx = 0,
): string | string[] => {
  const themeColor = color[colorQuery];
  if (Array.isArray(themeColor)) {
    return themeColor[idx];
  }
  return themeColor;
};
