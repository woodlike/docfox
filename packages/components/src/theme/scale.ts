export interface ScaledColorProps {
  color: string;
  idx: number;
}

export type Scale = string | number;
export type ScaleBox = [Scale, Scale, Scale, Scale];
export type ScaleArea = number | [number, number] | [number, number, number] | [number, number, number, number];

export function getThemeScale(area: ScaleArea, scale: Scale[]): Scale | Scale[] {
  return Array.isArray(area) ? area.map(value => scale[value]) : scale[area];
}

export function createBox(scale: Scale | Scale[]): ScaleBox {
  if (Array.isArray(scale)) {
    if (scale.length === 2) {
      return [...scale, ...scale] as ScaleBox;
    }
    if (scale.length === 3) {
      return [...scale, scale[1]] as ScaleBox;
    }
    return [...scale] as ScaleBox;
  }
  return [scale, scale, scale, scale];
}

export const create = (area: ScaleArea, scale: Scale[]): ScaleBox => createBox(getThemeScale(area, scale));

export const toCSSPixel = (box: ScaleBox): string => box.map(value => `${value}px`).join(' ');
export const toCSSString = (box: ScaleBox): string => box.map(value => `${value}`).join(' ');
