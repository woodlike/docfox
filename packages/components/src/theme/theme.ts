import { base } from '@theme-ui/presets';
import { theme as themeQuery, Colors, toRGB } from 'theme-query';

export interface Theme {
  readonly borderStyles: string[];
  readonly borderWidths: number[];
  readonly breakpoints: string[];
  readonly colors: ThemeColor & GenericThemeColor;
  readonly fonts: ThemeFontsProps;
  readonly fontSizes: number[];
  readonly space: number[];
}

export interface GenericThemeColor {
  readonly [key: string]: string | string[];
}

export interface ThemeFontsProps {
  readonly body: string;
  readonly heading: ThemeHeadingProps;
}

export interface ThemeHeadingProps {
  readonly display: string;
}

export interface ThemeColor {
  readonly primary: string;
  readonly secondary: string;
  readonly text: string;
  readonly border: string;
  readonly muted: string;
  readonly mutedHover: string;
  readonly headline: string;
  readonly link: string;
  readonly textInverted: string;
}

const borderStyles = [
  'none',
  'hidden',
  'dotted',
  'dashed',
  'solid',
  'double',
  'groove',
  'ridge',
  'inset',
  'outset',
];
const borderWidths = [1, 2, 3, 4];
const breakpoints = ['0px', '480px', '768px', '990px', '1024px', '1440px'];
const fontSizes: number[] = [12, 14, 16, 18, 20, 24, 32, 44, 72];
const fontWeights: number[] = [100, 200, 300, 400, 500, 600, 700, 800, 900];
const spaces = [0, 3, 6, 12, 18, 24, 30, 36, 48, 72];

const colors: Colors = {
  codeBg: toRGB([38, 42, 51]),
  corals: [toRGB([255, 113, 99]), toRGB([229, 85, 78])],
  whites: [
    toRGB([255, 255, 255]),
    toRGB([240, 246, 246]),
    toRGB([240, 240, 240]),
    toRGB([249, 246, 241]),
    toRGB([247, 247, 247]),
  ],
  grays: [
    toRGB([222, 223, 224]),
    toRGB([159, 188, 208]),
    toRGB([142, 149, 154]),
    toRGB([89, 103, 113]),
    toRGB([77, 77, 79]),
    toRGB([56, 72, 82]),
  ],
  blacks: [toRGB([34, 34, 34]), toRGB([51, 51, 51])],
  beiges: [toRGB([250, 245, 244])],
  sand: toRGB([204, 153, 102]),
  yellows: [toRGB([223, 199, 129]), toRGB([255, 202, 84])],
};
export const theme = {
  ...base,
  ...themeQuery,

  borderStyles,
  borderWidths,
  breakpoints,
  colors: {
    ...colors,
    background: colors.whites[0],
    border: colors.grays[0],
    headline: colors.grays[5],
    link: colors.grays[3],
    primary: colors.corals[0],
    muted: colors.grays[3],
    mutedHover: colors.grays[2],
    secondary: colors.corals[1],
    text: colors.blacks[0],
    textInverted: colors.whites[1],
  },
  fontSizes,
  fontWeights,
  fonts: {
    ...themeQuery.fonts,
    heading: {
      display: `"Museo", serif`,
      secondary: `"MuseoSans", Helvetica, sans-serif`,
      campaign: `"Challista_signature", serif`,
    },
    monospace: `"IBM Plex Mono", monospace`,
  },
  header: ['70px', '70px', '70px', '80px', '80px', '80px'],
  letterSpacings: [0.5, 1, 1.5],
  space: spaces,
  transition: {
    duration: [...themeQuery.transition.duration],
    timing: [
      ...themeQuery.transition.timing,
      'cubic-bezier(0.000, 0.755, 0.450, 0.910)',
      'cubic-bezier(0.770, 0.000, 0.175, 1.000)',
    ],
  },
};
