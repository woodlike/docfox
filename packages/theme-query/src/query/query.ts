/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { ThemeQueryConfig, ThemeQuery, ThemeProps } from '.';
import { splitFromDot } from '..';
import { SxStyleProp } from 'theme-ui';

export const query = (theme: ThemeProps, path: string) => {
  const result = [];
  for (const prop in theme) {
    if (prop === path) {
      result.push(theme[prop]);
      break;
    }
    const value = Object.keys(theme[prop]).find((val: string) =>
      path.includes('.') ? val === splitFromDot(path)[0] : val === path,
    );

    if (value && theme[prop][value]) {
      if (path.includes('.')) {
        const [key, val] = splitFromDot(path);
        result.push(theme[prop][key][val]);
        break;
      }
      result.push(theme[prop][value]);
      break;
    }
  }
  return result[0];
};

export function create(config: ThemeQueryConfig): ThemeQuery {
  const validConfig = config && config.theme && typeof config.theme === 'object';
  if (!validConfig) {
    throw Error('A valid configuration with a theme property must be provided');
  }
  const { theme, styles } = config;

  return (objPath: string) => {
    const queryResult = query(theme, objPath);

    if (Array.isArray(queryResult) && queryResult.length > 0 && styles === 'object') {
      return (idx: number | 'all'): SxStyleProp | SxStyleProp[] => {
        return Number.isInteger(idx as number) ? queryResult[idx as number] : queryResult;
      };
    }
    return styles === 'object' ? queryResult : `${queryResult}`;
  };
}
