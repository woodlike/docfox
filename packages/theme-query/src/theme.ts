import {
  borderWidths,
  colors,
  fontFamily,
  fontSizes,
  fontWeights,
  letterSpacings,
  spaces,
  timing,
  duration,
} from './tokens';

export const theme = {
  borderWidths,
  colors,
  fonts: {
    ...fontFamily,
  },
  fontSizes,
  fontWeights,
  letterSpacings,
  spaces,
  transition: {
    duration,
    timing,
  },
};
export default theme;
