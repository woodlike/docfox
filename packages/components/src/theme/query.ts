import { create, ThemeQuery, ThemeQueryConfig } from 'theme-query';
import { useThemeUI } from 'theme-ui';
import { theme } from './theme';

export const qt = create({ theme, styles: 'object' });

export interface ThemeQueryHook {
  qt: ThemeQuery;
}

export function useThemeQuery(): ThemeQueryHook {
  const { theme: themeUI } = useThemeUI();
  const config: ThemeQueryConfig = {
    theme: themeUI ? themeUI : theme,
    styles: 'object',
  };
  return { qt: create(config) };
}
