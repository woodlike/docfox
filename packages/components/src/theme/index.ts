import * as ColorQuery from './color';
import * as InternalScale from './scale';

export * from './query';
export * from './scale';
export * from './theme';

export const Color = { ...ColorQuery };
export const Scale = { ...InternalScale };
