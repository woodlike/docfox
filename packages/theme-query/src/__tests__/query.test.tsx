/** @jsx jsx */
import { jsx } from 'theme-ui';
import { render } from '@testing-library/react';

import * as ThemeQuery from '../query';
import { ThemeQueryConfig } from '../query';
import { theme } from '../theme';

describe('StyledFactory', () => {
  let sandMock: unknown;
  let coralMock: unknown;
  let coralDarkMock: unknown;
  let headingDefault: unknown;
  let headingSecondary: unknown;

  beforeAll(() => {
    sandMock = 'rgb(204, 153, 102)';
    coralMock = 'rgb(255, 113, 99)';
    coralDarkMock = 'rgb(229, 85, 78)';
    headingDefault = '"Museo", serif';
    headingSecondary = `"Museo500", serif`;
  });

  afterAll(() => {
    sandMock = undefined;
    coralMock = undefined;
    coralDarkMock = undefined;
    headingDefault = undefined;
    headingSecondary = undefined;
  });

  describe('StyledFactory.create(config)', () => {
    it('should throw an error if no configuration is provided', () => {
      const errMsg = 'A valid configuration with a theme property must be provided';

      expect(() => ThemeQuery.create((null as unknown) as ThemeQueryConfig)).toThrowError(errMsg);
      expect(() => ThemeQuery.create((undefined as unknown) as ThemeQueryConfig)).toThrowError(errMsg);
      expect(() => ThemeQuery.create((1 as unknown) as ThemeQueryConfig)).toThrowError(errMsg);
      expect(() => ThemeQuery.create(('hi' as unknown) as ThemeQueryConfig)).toThrowError(errMsg);
    });

    it('should throw an error if the theme is not provided', () => {
      const errMsg = 'A valid configuration with a theme property must be provided';
      expect(() =>
        ThemeQuery.create({
          theme: (undefined as unknown) as {},
          styles: 'object',
        }),
      ).toThrowError(errMsg);
      expect(() => ThemeQuery.create({ theme: (null as unknown) as {}, styles: 'object' })).toThrowError(errMsg);
      expect(() => ThemeQuery.create({ theme: (1 as unknown) as {}, styles: 'object' })).toThrowError(errMsg);
      expect(() => ThemeQuery.create({ theme: ('' as unknown) as {}, styles: 'object' })).toThrowError(errMsg);
    });

    it('should return a qt function on initialization', () => {
      const qt = ThemeQuery.create({ theme, styles: 'object' });
      expect(qt).toStrictEqual(expect.any(Function));
    });
  });

  describe('qt(...)', () => {
    it('should return all the defined theme colors', () => {
      const qt = ThemeQuery.create({ theme, styles: 'object' });
      expect(qt('colors')).toMatchSnapshot();
    });

    it('should return all the corals colors which is a nested color array', () => {
      const qt = ThemeQuery.create({ theme, styles: 'object' });

      expect(qt('corals')(0)).toBe(coralMock);
      expect(qt('corals')(1)).toBe(coralDarkMock);
    });

    it('should return the sand color which is a flat color', () => {
      const qt = ThemeQuery.create({ theme, styles: 'object' });

      expect(qt('sand')).toBe(sandMock);
    });

    it('should return the value from fonts.heading.display ', () => {
      const qt = ThemeQuery.create({ theme, styles: 'object' });

      expect(qt('heading.display')).toBe(headingDefault);
    });

    it('should return the value from fonts.heading.secondary ', () => {
      const qt = ThemeQuery.create({ theme, styles: 'object' });

      expect(qt('heading.secondary')).toBe(headingSecondary);
    });

    it('should return the complete array from the property value', () => {
      const qt = ThemeQuery.create({ theme, styles: 'object' });

      expect(qt('corals')('all')).toStrictEqual(theme.colors.corals);
      expect(qt('borderWidths')('all')).toStrictEqual(theme.borderWidths);
    });
  });

  describe('Component Styling', () => {
    it('should receive the necessary styling for the component', () => {
      const qt = ThemeQuery.create({ theme, styles: 'object' });

      const { getByTestId, unmount } = render(
        <div
          data-testid="test-el"
          sx={{
            fontSize: qt('fontSizes')(4),
            fontFamily: qt('heading.secondary'),
            color: qt('sand'),
            bg: qt('corals')(0),
          }}>
          Hi Mom!!!
        </div>,
      );

      const div = getByTestId('test-el');
      const styles = getComputedStyle(div);
      expect(styles.getPropertyValue('color')).toBe(sandMock);
      expect(styles.getPropertyValue('background-color')).toBe(coralMock);
      expect(styles.getPropertyValue('font-size')).toBe('20px');
      unmount();
    });
  });
});
