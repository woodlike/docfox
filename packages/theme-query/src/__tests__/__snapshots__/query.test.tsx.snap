// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyledFactory qt(...) should return all the defined theme colors 1`] = `
Object {
  "beiges": Array [
    "rgb(250, 245, 244)",
  ],
  "blacks": Array [
    "rgb(34, 34, 34)",
    "rgb(51, 51, 51)",
  ],
  "corals": Array [
    "rgb(255, 113, 99)",
    "rgb(229, 85, 78)",
  ],
  "grays": Array [
    "rgb(222, 223, 224)",
    "rgb(159, 188, 208)",
    "rgb(142, 149, 154)",
    "rgb(89, 103, 113)",
    "rgb(77, 77, 79)",
    "rgb(56, 72, 82)",
  ],
  "sand": "rgb(204, 153, 102)",
  "whites": Array [
    "rgb(255, 255, 255)",
    "rgb(240, 246, 246)",
    "rgb(240, 240, 240)",
    "rgb(249, 246, 241)",
    "rgb(247, 247, 247)",
  ],
  "yellows": Array [
    "rgb(223, 199, 129)",
    "rgb(255, 202, 84)",
  ],
}
`;
